### PR TITLE
Task av/j34 create podcast page

### PR DESCRIPTION
**Pull request by artem-totality**

### CODE REVIEW

File: frontend/src/components/common/button/button.tsx [hunk 1]
```diff
@@ -1,10 +1,12 @@
 import { ButtonColor, ButtonStyle, ButtonType } from 'common/enums/enums';
 import { getAllowedClasses } from 'helpers/dom/dom';
+import React from 'react';
```
Comment (ghost) [inline frontend/src/components/common/button/button.tsx]:
```suggestion
```

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 2]
```diff
@@ -0,0 +1,80 @@
+import { PodcastCreatePayload } from 'common/types/types';
+import { DataStatus } from 'common/enums/enums';
+import { podcast as podcastActions } from 'store/actions';
+import { useAppSelector, useDispatch, useState } from 'hooks/hooks';
+import { Input, Button } from 'components/common/common';
+import styles from './styles.module.scss';
+
+const CreatePodcast: React.FC = () => {
+  const [podcastCreatePayload, setPodcastCreatePayload] =
+    useState<PodcastCreatePayload>({
```
Comment (ghost) [inline frontend/src/components/create-podcast/create-podcast.tsx]:
```suggestion
    useState<PodcastCreatePayload>({
```
use useForm hook for all forms

File: frontend/src/services/podcast-api/podcast-api.service.ts [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+class PodcastApi {
+  #http: Http;
+  #apiPrefix: string;
+
+  constructor({ http, apiPrefix }: Constructor) {
+    this.#http = http;
+    this.#apiPrefix = apiPrefix;
+  }
+
+  public postPodcast(payload: PodcastCreatePayload): Promise<Podcast> {
```
Comment (ghost) [inline frontend/src/services/podcast-api/podcast-api.service.ts]:
```suggestion
  public create(payload: PodcastCreatePayload): Promise<Podcast> {
```

File: frontend/src/store/podcast/actions.ts [hunk 4]
```diff
@@ -0,0 +1,18 @@
+import { createAsyncThunk } from '@reduxjs/toolkit';
+import {
+  AsyncThunkConfig,
+  Podcast,
+  PodcastCreatePayload,
+} from 'common/types/types';
+import { ActionType } from './common';
+
+// prettier-ignore
```
Comment (ghost) [inline frontend/src/store/podcast/actions.ts]:
```suggestion
```

File: frontend/src/store/podcast/reducer.ts [hunk 5]
```diff
@@ -0,0 +1,28 @@
+import { createReducer } from '@reduxjs/toolkit';
```
Comment (ghost) [inline frontend/src/store/podcast/reducer.ts]: your page is responsible for creating or editing a podcast. Therefore, the reducer and actions should be called `configurate-pocast`

File: frontend/src/store/root-reducer.ts [hunk 6]
```diff
@@ -1,8 +1,10 @@
 import { reducer as auth } from './auth/reducer';
+import { reducer as podcast } from './podcast/reducer';
```
Comment (ghost) [inline frontend/src/store/root-reducer.ts]:
```suggestion
import { reducer as configuratePodcast } from './podcast/reducer';
```

File: frontend/src/components/common/button/button.tsx [hunk 7]
```diff
@@ -14,6 +16,9 @@ const Button: React.FC<Props> = ({
   type = ButtonType.BUTTON,
   buttonStyle = ButtonStyle.PRIMARY,
   buttonColor = ButtonColor.PINK,
+  onClick = (): void => {
+    return;
+  },
```
Comment (ghost) [inline frontend/src/components/common/button/button.tsx]:
```suggestion
  onClick
```

**Review by ghost — CHANGES_REQUESTED** [review]

**Review by ghost — CHANGES_REQUESTED** [review]
don't forget to add this page to app routing

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+      name: '',
+      userId: 7,
+    });
+
+  const { createPodcastStatus } = useAppSelector(({ podcast }) => ({
+    createPodcastStatus: podcast.dataStatus,
+  }));
+  const dispatch = useDispatch();
+
+  const isFormDisable = createPodcastStatus === DataStatus.PENDING;
```
Comment (yevhenprybytok) [inline frontend/src/components/create-podcast/create-podcast.tsx]:
```suggestion
  const isFormDisabled = createPodcastStatus === DataStatus.PENDING;
```

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+                value={podcastCreatePayload.name}
+                name="name"
+                isDisabled={isFormDisable}
+                onChange={handleChange}
+              />
+            </p>
+            <p>
+              <Input
+                label="Description"
+                value={''}
```
Comment (yevhenprybytok) [inline frontend/src/components/create-podcast/create-podcast.tsx]:
```suggestion
                value=""
```

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 10]
```diff
@@ -0,0 +1,80 @@
...
+                label="Genre"
+                value={''}
+                name="genre"
+                isDisabled={isFormDisable}
+                onChange={handleChange}
+              />
+            </p>
+          </div>
+          <div>
+            <img src={'https://via.placeholder.com/150'} />
```
Comment (yevhenprybytok) [inline frontend/src/components/create-podcast/create-podcast.tsx]: Fallback images should be located in assets directory

**Review by yevhenprybytok — CHANGES_REQUESTED** [review]

File: frontend/src/common/enums/app/app-route.enum.ts [hunk 11]
```diff
@@ -2,6 +2,7 @@ enum AppRoute {
   ROOT = '/',
   SIGN_IN = '/sign-in',
   SIGN_UP = '/sign-up',
+  CREATE_PODCAST = '/create-podcast',
```
Comment (ghost) [inline frontend/src/common/enums/app/app-route.enum.ts]:
```suggestion
  EDIT_PODCAST = '/podcasts/edit/:id',
```
we'll use this page for creating & editing
(check documentation)

File: shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts [hunk 12]
```diff
@@ -0,0 +1,10 @@
+import * as Joi from 'joi';
+import { PodcastCreatePayloadKey } from '~/common/enums/enums';
+import { PodcastCreatePayload } from '~/common/types/types';
+
+const createPodcastSchema = Joi.object<PodcastCreatePayload>({
+  [PodcastCreatePayloadKey.NAME]: Joi.string().min(1).max(20).required(),
```
Comment (ghost) [inline shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts]: Create enum for:
rules - PodcastValidationRule
messages: PodcastValidationMessage
in the `shared/src/common/enums/validation/`

File: shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts [hunk 13]
```diff
@@ -0,0 +1,10 @@
+import * as Joi from 'joi';
+import { PodcastCreatePayloadKey } from '~/common/enums/enums';
+import { PodcastCreatePayload } from '~/common/types/types';
+
+const createPodcastSchema = Joi.object<PodcastCreatePayload>({
```
Comment (ghost) [inline shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts]:
```suggestion
const podcast = Joi.object<PodcastCreatePayload>({
```

Check the A3 criteria

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 14]
```diff
@@ -0,0 +1,69 @@
+import { useForm, FieldValues } from 'react-hook-form';
+import { PodcastCreatePayload } from 'common/types/types';
+import { joiResolver } from '@hookform/resolvers/joi';
+import { ButtonType, DataStatus } from 'common/enums/enums';
+import { podcast as podcastActions } from 'store/actions';
+import { useAppSelector, useDispatch } from 'hooks/hooks';
+import { Input, Button } from 'components/common/common';
+import styles from './styles.module.scss';
+import { PodcastCreatePayloadKey } from 'jabber-shared/common/enums/enums';
+import { createPodcastSchema } from 'validation-schemas/validation-schemas';
```
Comment (ghost) [inline frontend/src/components/create-podcast/create-podcast.tsx]:
```suggestion
import { podcast as podcastSchema } from 'validation-schemas/validation-schemas';
```

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 15]
```diff
@@ -0,0 +1,69 @@
+import { useForm, FieldValues } from 'react-hook-form';
+import { PodcastCreatePayload } from 'common/types/types';
+import { joiResolver } from '@hookform/resolvers/joi';
+import { ButtonType, DataStatus } from 'common/enums/enums';
+import { podcast as podcastActions } from 'store/actions';
+import { useAppSelector, useDispatch } from 'hooks/hooks';
+import { Input, Button } from 'components/common/common';
+import styles from './styles.module.scss';
+import { PodcastCreatePayloadKey } from 'jabber-shared/common/enums/enums';
```
Comment (ghost) [inline frontend/src/components/create-podcast/create-podcast.tsx]: don't import directly from shared packages

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 16]
```diff
@@ -0,0 +1,69 @@
+import { useForm, FieldValues } from 'react-hook-form';
```
Comment (ghost) [inline frontend/src/components/create-podcast/create-podcast.tsx]: move useForm for `frontend/src/hooks`

File: frontend/src/store/configurate-pocast/common.ts [hunk 17]
```diff
@@ -0,0 +1,5 @@
+enum ActionType {
+  POST_PODCAST = 'podcast/post-podcast',
```
Comment (ghost) [inline frontend/src/store/configurate-pocast/common.ts]:
```suggestion
  POST_PODCAST = 'configurate/create-podcast',
```

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 18]
```diff
@@ -0,0 +1,69 @@
...
+
+const CreatePodcast: React.FC = () => {
+  const {
+    control,
+    handleSubmit,
+    formState: { errors },
+  } = useForm<FieldValues>({
+    defaultValues: {
+      [PodcastCreatePayloadKey.NAME]: '',
+      [PodcastCreatePayloadKey.USER_ID]: 7,
```
Comment (ghost) [inline frontend/src/components/create-podcast/create-podcast.tsx]: move it to constants
frontend/src/components/create-podcast/common/constants.ts

**Review by ghost — CHANGES_REQUESTED** [review]

File: frontend/src/common/enums/app/app-route.enum.ts [hunk 19]
```diff
@@ -2,6 +2,7 @@ enum AppRoute {
   ROOT = '/',
   SIGN_IN = '/sign-in',
   SIGN_UP = '/sign-up',
+  EDIT_PODCAST = '/podcasts/edit/:id',
```
Comment (what1s1ove) [inline frontend/src/common/enums/app/app-route.enum.ts]:
```suggestion
  POCAST_EDIT_$ID = '/podcasts/edit/:id?',
```
Add `$` for dynamic params

File: frontend/src/components/app/app.tsx [hunk 20]
```diff
@@ -7,6 +7,7 @@ import {
   Header,
   NotFound,
 } from 'components/common/common';
+import CreatePodcast from 'components/create-podcast/create-podcast';
```
Comment (what1s1ove) [inline frontend/src/components/app/app.tsx]:
```suggestion
import ConfiguratePodcast from 'components/configurate-podcast/configurate-podcast';
```

File: frontend/src/components/create-podcast/common/constants.ts [hunk 21]
```diff
@@ -0,0 +1,9 @@
+import { PodcastCreatePayloadKey } from 'common/enums/enums';
+import { PodcastCreatePayload } from 'common/types/types';
+
+const DEFAULT_PODCAST_PAYLOAD: PodcastCreatePayload = {
+  [PodcastCreatePayloadKey.NAME]: '',
+  [PodcastCreatePayloadKey.USER_ID]: '1',
```
Comment (what1s1ove) [inline frontend/src/components/create-podcast/common/constants.ts]:
```suggestion
  // shared/src/common/constants/user.ts
  const DEFAULT_USER_ID = -1;

  [PodcastCreatePayloadKey.USER_ID]: DEFAULT_USER_ID,
```

File: shared/src/common/types/podcast/podcast-create-payload.type.ts [hunk 22]
```diff
@@ -2,7 +2,7 @@ import { PodcastCreatePayloadKey } from '~/common/enums/enums';
 
 type PodcastCreatePayload = {
   [PodcastCreatePayloadKey.NAME]: string;
-  [PodcastCreatePayloadKey.USER_ID]: number;
+  [PodcastCreatePayloadKey.USER_ID]: string;
```
Comment (what1s1ove) [inline shared/src/common/types/podcast/podcast-create-payload.type.ts]:
```suggestion
  [PodcastCreatePayloadKey.USER_ID]: number;
```

File: shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts [hunk 23]
```diff
@@ -0,0 +1,30 @@
...
+    .max(PodcastValidationRule.PODCAST_NAME_MAX_LENGTH)
+    .required()
+    .messages({
+      'string.empty': PodcastValidationMessage.PODCAST_NAME_REQUIRE,
+      'string.min': PodcastValidationMessage.PODCAST_NAME_MIN_LENGTH,
+      'string.max': PodcastValidationMessage.PODCAST_NAME_MAX_LENGTH,
+    }),
+  [PodcastCreatePayloadKey.USER_ID]: Joi.number()
+    .integer()
+    .positive()
```
Comment (what1s1ove) [inline shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts]:
```suggestion
```

File: shared/src/validation-schemas/podcast/podcast.ts [hunk 24]
```diff
@@ -0,0 +1 @@
+export * from './create-podcast.validation-schema';
```
Comment (what1s1ove) [inline shared/src/validation-schemas/podcast/podcast.ts]:
```suggestion
export { podcast } from './create-podcast.validation-schema';
```

File: frontend/src/store/configurate-pocast/common.ts [hunk 25]
```diff
@@ -0,0 +1,5 @@
+enum ActionType {
+  POST_PODCAST = 'configurate/create-podcast',
```
Comment (what1s1ove) [inline frontend/src/store/configurate-pocast/common.ts]:
```suggestion
  CREATE_PODCAST = 'configurate-podcast/create-podcast',
```

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+  });
+
+  const { createPodcastStatus } = useAppSelector(({ configuratePodcast }) => ({
+    createPodcastStatus: configuratePodcast.dataStatus,
+  }));
+  const dispatch = useDispatch();
+
+  const isFormDisabled = createPodcastStatus === DataStatus.PENDING;
+
+  const onSubmit = (data: PodcastCreatePayload): void => {
```
Comment (what1s1ove) [inline frontend/src/components/create-podcast/create-podcast.tsx]: check **A9** criteria

File: frontend/src/components/create-podcast/create-podcast.tsx [hunk 27]
```diff
@@ -0,0 +1,59 @@
...
+    dispatch(podcastActions.create(data));
+  };
+
+  return (
+    <div>
+      <form
```
Comment (what1s1ove) [inline frontend/src/components/create-podcast/create-podcast.tsx]: move form to `frontend/src/components/create-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx`

**Review by what1s1ove — CHANGES_REQUESTED** [review]

File: shared/src/common/constants/user.ts [hunk 28]
```diff
@@ -0,0 +1,3 @@
+const DEFAULT_USER_ID = -1;
```
Comment (liza-veis) [inline shared/src/common/constants/user.ts]: Move it to user folder

File: shared/src/common/enums/validation/podcast-validation-message.enum.ts [hunk 29]
```diff
@@ -0,0 +1,11 @@
+import { PodcastValidationRule } from './podcast-validation-rule.enum';
+
+const PodcastValidationMessage = {
+  PODCAST_NAME_REQUIRE: 'Podcast name is required',
+  PODCAST_NAME_MIN_LENGTH: `Podcast name must be at least ${PodcastValidationRule.PODCAST_NAME_MIN_LENGTH} character long`,
+  PODCAST_NAME_MAX_LENGTH: `Podcast name must be at most ${PodcastValidationRule.PODCAST_NAME_MAX_LENGTH} characters long`,
+  USER_ID_REQUIRE: 'User id required',
+  USER_ID_NUMBER_FORMAT: 'User id must be integer number',
```
Comment (liza-veis) [inline shared/src/common/enums/validation/podcast-validation-message.enum.ts]:
```suggestion
  USER_ID_NUMBER_FORMAT: 'User id must be an integer',
```

File: frontend/src/common/constants/constants.ts [hunk 30]
```diff
@@ -0,0 +1 @@
+export { DEFAULT_USER_ID } from 'jabber-shared/common/constants/constants';
```
Comment (liza-veis) [inline frontend/src/common/constants/constants.ts]: Move it to the user folder "frontend/src/common/constants/user" and then re-export from this file

**Review by liza-veis — CHANGES_REQUESTED** [review]

**Comment by liza-veis** [comment]
Why do you often put something in the "p" tag? If you want to add paddings you can just use styles or is there another reason?

**Comment by artem-totality** [comment]
> Why do you often put something in the "p" tag? If you want to add paddings you can just use styles or is there another reason?

It was the simplest way to make this padding in this situation without redundant styles.

**Review by liza-veis — DISMISSED** [review]

File: shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts [hunk 31]
```diff
@@ -0,0 +1,28 @@
+import * as Joi from 'joi';
```
Comment (ghost) [inline shared/src/validation-schemas/podcast/create-podcast.validation-schema.ts]: rename file to ``shared/src/validation-schemas/podcast/podcast.validation-schema.ts

File: frontend/src/components/configurate-podcast/configurate-podcast.tsx [hunk 32]
```diff
@@ -0,0 +1,16 @@
...
+import { podcast as podcastActions } from 'store/actions';
+import { useDispatch } from 'hooks/hooks';
+import { ConfiguratePodcastForm } from './components/components';
+
+const ConfiguratePodcast: React.FC = () => {
+  const dispatch = useDispatch();
+
+  const handleCreatePodcast = (data: PodcastCreatePayload): void => {
+    dispatch(podcastActions.create(data));
+  };
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/configurate-podcast.tsx]:
```suggestion
  const handleCreatePodcast = (payload: PodcastCreatePayload): void => {
    dispatch(podcastActions.create(payload));
  };
```

File: frontend/src/components/configurate-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx [hunk 33]
```diff
@@ -0,0 +1,55 @@
...
+                name="name"
+                control={control}
+                errors={errors}
+              />
+            </p>
+          </fieldset>
+        </div>
+        <div>
+          <img src={logoCut} />
+        </div>
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx]:
```suggestion
```

File: frontend/src/components/configurate-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx [hunk 34]
```diff
@@ -0,0 +1,55 @@
...
+      </div>
+      <p>
+        <Button label="Save" type={ButtonType.SUBMIT} />
+      </p>
```
Comment (ghost) [inline frontend/src/components/configurate-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx]:
```suggestion
        <Button label="Save" type={ButtonType.SUBMIT} />
```

Comment (ghost) [reply frontend/src/components/configurate-podcast/components/configurate-podcast-form/configurate-podcast-form.tsx]: also, move button to fieldset (to disable it too while saving)

**Review by ghost — APPROVED** [review]

**Pull request merged**